Drop /var/lib/cockpit

martinpitt opened the pull request:
Spotted here: https://github.com/cockpit-project/cockpit/pull/14578#discussion_r485512928

allisonkarlitskaya commented on src/ssh/cockpitsshrelay.c:
```diff
@@ -477,35 +436,39 @@ write_tmp_knownhosts_file (CockpitSshData *data,
 
-  fp = fopen (tmp_knownhost_file, "a");
-  if (fp == NULL)
+  for (;;)
```
I think maybe you planned to try multiple `write()` in case of a partial result, but you never actually implemented this (instead opting to give an error message in this case).  This loop only ever runs once, as a result.

Calling `posix_fallocate()` first is a pretty reasonable compromise here (and generally improves results on extents-based filesystems).

martinpitt replied:
The loop was just meant to retry on `EINTR`. But indeed that writes a file and may get a short write; I think this is not really worth the trouble, and I'll just move this to `g_file_set_contents()`. This has absolutely no performance requirements, and once `create_knownhosts_temp()` succeeds, we can close the fd and reopen the file.

allisonkarlitskaya reviewed (changes requested):
Good start.

I did a bit of grepping around and also found at least these:

```
containers/ws/atomic-install:mkdir -p /host/var/lib/cockpit
containers/ws/atomic-install:chmod 775 /host/var/lib/cockpit
containers/ws/atomic-install:chown root:wheel /host/var/lib/cockpit
```

and

```
pkg/lib/machines.js:        known_hosts_path = "/var/lib/cockpit/known_hosts";
```

and

(probably a very old)
```
test/verify/check-login:        self.allow_journal_messages(".*/var/lib/cockpit/machines.json.*Permission denied")
```

and also `cockpit-ws.dirs` in the debian packaging.

martinpitt commented:
@allisonkarlitskaya : Good point about the unexpected message and the container directories, I cleaned them up. Also rewrote/simplified the temporary known hosts writing.

I kept pkg/lib/machines.js as this is explicitly fallback code for running against *old* cockpit-ws versions. Cleaning that up is independent from the changes here, and first needs  a little bit of research if we still need to support these old versions. For now I just made a TODO item for this.

allisonkarlitskaya commented on src/ssh/cockpitsshrelay.c:
```diff
@@ -477,35 +436,29 @@ write_tmp_knownhosts_file (CockpitSshData *data,
       *problem = "internal-error";
       return FALSE;
     }
+  close (fd);
```
Maybe add a comment here about now that the `O_EXCL`-created file exists with the given name, the security-sensitive work of `g_file_open_tmp()` is done, and we're free to use the name below?

martinpitt replied:
Added.

allisonkarlitskaya commented on src/ssh/cockpitsshrelay.c:
```diff
@@ -477,35 +436,29 @@ write_tmp_knownhosts_file (CockpitSshData *data,
-                 data->logname, g_strerror (errno));
-      fclose (fp);
+      g_warning ("%s: couldn't write data to temporary known host file %s: %s", data->logname, tmp_knownhost_file, error->message);
+      close (fd);
```
You already closed this above.

martinpitt replied:
:facepalm: fixed

allisonkarlitskaya reviewed (changes requested):
Very nice!  Only a couple of minor nags.

allisonkarlitskaya reviewed (approved):
Thanks!
